tests/unit/test_charm.py

```diff
@@ -1,35 +1,88 @@
 from unittest.mock import Mock, call, mock_open, patch
 from uuid import uuid4
 
 from jinja2 import Template
 from ops.model import ActiveStatus, BlockedStatus
 from ops.testing import Harness
 
-from charm import NginxCharm
+from charm import NginxCharm, file_hash, install_ca_cert, write_file
 
-DEFAULT_CONFIG = {"host": str(uuid4()), "port": random.randint(10, 20)}
+DEFAULT_CONFIG = {
+    "host": str(uuid4()),
+    "port": random.randint(10, 20),
+    "ssl_cert": "dGVzdF9jZXJ0==",  # Padded base64 strings
+    "ssl_key": "dGVzdF9rZXk=",  # Padded base64 strings
+    "ssl_ca": "dGVzdF9jYV9jZXJ0==",
+}
+NEW_CONFIG = {
+    "host": str(uuid4()),
+    "port": random.randint(10, 20),
+    "ssl_cert": "dGVzdF9jZXJ0==",  # Padded base64 strings
+    "ssl_key": "dGVzdF9rZXk=",  # Padded base64 strings
+    "ssl_ca": "dGVzdF9jYV9jZXJ0==",
+}
 STORED_CONFIG = {"host": str(uuid4()), "port": random.randint(10, 20), "publishes": {}}
 
 
 class TestCharm(unittest.TestCase):
-    def test_config_changed(self):
+    @patch("charm.install_ca_cert")
+    @patch("charm.write_file")
+    @patch("os.path.exists", return_value=False)
+    @patch("os.makedirs")
+    @patch("os.chown")
+    @patch("os.chmod")
+    @patch("subprocess.check_call")
+    def test_config_changed(
+        self,
+        mock_check_call,
+        mock_chmod,
+        mock_chown,
+        mock_makedirs,
+        mock_path_exists,
+        mock_write_file,
+        mock_install_ca_cert,
+    ):
         harness = Harness(NginxCharm)
         self.addCleanup(harness.cleanup)
         harness.begin()
         harness.charm._render_config = Mock()
         harness.charm._reload_config = Mock()
-        # default_config = DEFAULT_CONFIG
         harness.update_config(DEFAULT_CONFIG)
-        default_config = {**DEFAULT_CONFIG, "publishes": {}}
+        harness.update_config(NEW_CONFIG)
+        default_config = {**NEW_CONFIG, "publishes": {}, "ssl_enabled": True}
+        print("actual config:", harness.charm._stored.config)
+        print("expected config:", default_config)
         self.assertEqual(harness.charm._stored.config, default_config)
         self.assertTrue(harness.charm._render_config.called)
         self.assertTrue(harness.charm._reload_config.called)
+        mock_makedirs.assert_called_with("/etc/nginx/ssl", 0o755)
+        mock_chown.assert_called_with("/etc/nginx/ssl", 0, 0)
+        mock_chmod.assert_called_with("/etc/nginx/ssl", 0o755)
+        mock_write_file.assert_any_call(
+            "/etc/nginx/ssl/server.crt",
+            b64decode("dGVzdF9jZXJ0=="),
+            "root",
+            "root",
+            0o644,
+        )
+        mock_write_file.assert_any_call(
+            "/etc/nginx/ssl/server.key",
+            b64decode("dGVzdF9rZXk="),
+            "root",
+            "root",
+            0o640,
+        )
+        # mock_install_ca_cert.assert_called_with(
+        #     b64decode("dGVzdF9jYV9jZXJ0"), "nginx-server.crt"
+        # )
```

Comment:
why do you need to comment this out here but not in #15 ?

Reply:
This is due to the conflict of ssl_ca and ca_cert. In the charm code this is called only if "ca_cert" exists in the config, but in unit test I have to use ssl_ca to pass other tests. I comment this out for a temporary solution. In https://github.com/canonical/charm-nginx/pull/15 I have modified both to use the same name so the issue doesn't exist.

Reply:
Actually should we remove unit tests related to TLS, since they will need to be modified further and the CI tests are passing even without 100% coverage.

Reply:
either way. I would personally change ca_cert to ssl_ca, uncomment these lines, and iterate later

Reply:
Just applied the change.